text/0730-semver-for-ts.md

```diff
@@ -0,0 +1,912 @@
+
+[RFC #0724]: https://github.com/emberjs/rfcs/pull/724
+
+
+### Outline <!-- omit in toc -->
+
+- [Motivation](#motivation)
+- [Detailed design](#detailed-design)
+  - [Background: TypeScript and Semantic Versioning](#background-typescript-and-semantic-versioning)
+  - [Defining breaking changes](#defining-breaking-changes)
+    - [Supported compiler versions](#supported-compiler-versions)
+      - [Simple majors](#simple-majors)
+      - [Rolling support windows](#rolling-support-windows)
+    - [Definitions](#definitions)
+    - [Breaking changes](#breaking-changes)
+      - [Symbols](#symbols)
+      - [Interfaces, Type Aliases, and Classes](#interfaces-type-aliases-and-classes)
+      - [Functions](#functions)
+    - [Non-breaking changes](#non-breaking-changes)
+      - [Symbols](#symbols-1)
+      - [Interfaces, Type Aliases, and Classes](#interfaces-type-aliases-and-classes-1)
+      - [Functions](#functions-1)
+    - [Bug fixes](#bug-fixes)
+    - [Strictness](#strictness)
+    - [Module interop](#module-interop)
+  - [Design summary](#design-summary)
+  - [Conformance](#conformance)
+- [How we teach this](#how-we-teach-this)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [No policy](#no-policy)
+  - [Decouple TypeScript support from LTS cycles](#decouple-typescript-support-from-lts-cycles)
+- [Unresolved questions](#unresolved-questions)
+- [Appendices](#appendices)
+  - [Appendix A: Existing Implementations](#appendix-a-existing-implementations)
+  - [Appendix B: Tooling](#appendix-b-tooling)
+    - [Documenting supported versions and policy](#documenting-supported-versions-and-policy)
+    - [Detect breaking changes in types](#detect-breaking-changes-in-types)
+    - [Mitigate breaking changes](#mitigate-breaking-changes)
+      - [Avoiding user constructability](#avoiding-user-constructability)
+      - [Updating types to maintain compatibility](#updating-types-to-maintain-compatibility)
+      - ["Downleveling" types](#downleveling-types)
+      - [Opt-in future types](#opt-in-future-types)
+    - [Matching exports to public API](#matching-exports-to-public-api)
+
+
+## Motivation
+
+For TypeScript packages be good citizens of the broader, semantically-versioned JavaScript ecosystem, package authors need a useful definition of SemVer for TypeScript’s type system.
+
+This is somewhat more complicated than in other languages, even those other statically-typed languages with language-level SemVer guarantees (such as [Rust][rust-semver] and Elm), because TypeScript has an unusually flexible type system. In particular, its [structural type system][structural] means many more kinds of both breaking and non-breaking changes are possible than in languages with a [nominal type system][nominal]. Accordingly, this document proposes a definition of SemVer which accounts for the extra flexibility afforded by these features.
```

Comment:
Elm’s record types are structural IIUC. See for example the discussion of pattern matching in https://elm-lang.org/docs/records. I wonder how changes to them are treated by elm-package when automatically deciding the next package version?

Reply:
I had the same thought about Elm after I published this, and I may actually go poke at the relevant section of the build tooling to answer the question, as I would like to know! Unfortunately, the documentation of what Elm does [has gone missing](https://discourse.elm-lang.org/t/where-did-the-reference-for-package-semver-go/7122), and the info which existed there historically was [less detailed than I remembered](https://github.com/elm-lang/elm-package/commit/2076638d812d9dea641e6e695951495f523bb98f#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L26-L48).

Reply:
The implementation looks to be at https://github.com/elm/compiler/blob/770071accf791e8171440709effe71e78a9ab37c/builder/src/Deps/Diff.hs#L128-L136. I think any change in the number of fields is treated as a major change, while there’s some notion of “compatible renamings”, but I’m not sure.

Reply:
Just integrated the notes here at 3a353db.

As an amusing aside for lookers-on: @jeremy-w and I are in a shared community online and we both independently got to the above link at about the same time and linked it to each other independently back in March. 😅 Love it! 🤜🏼🤛🏼